### PR TITLE
Feature/ticket11 disallow dual provider roles

### PR DESCRIPTION
**Pull request by david415**
implements whatever this ticket says: https://github.com/katzenpost/katzenpost/issues/11

### CODE REVIEW

**Comment by mixmasala** [comment]
```
06:21:12.744 DEBU minclient/conn:W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1: PKI fetch successful.
06:21:12.745 DEBU minclient/conn:W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1: Descriptor for epoch 116527: &{Name:provider1 IdentityKey:2krwfNDfbakZCSTUUZYKXwdduzlEgS9Jfwm7eyZ0sCg= LinkKey:bAAHyaf/fl682PBj2VWzGRMVfWBV5t+oBaaKUg+6ZhE= MixKeys:map[116527:zHRcV2dookTuQcImGMt8MSBVBXMUwLR/uM3IXmjoikU= 116528:VWqAJ3pm0KmQnyCBvq0qJvp9NVdubU+m5szwpe6vkAk= 116529:BgJ01aJqj0OoqVBukmNPPJmfaF5LpHgTOgaeV9h0iys=] Addresses:map[tcp4:[172.28.1.3:59484]] Kaetzchen:map[echo:map[endpoint:echo] panda:map[endpoint:+panda] spool:map[endpoint:+spool]] Layer:255 LoadWeight:0 AuthenticationType:tofu}
06:21:12.746 DEBU minclient/conn:W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1: Dialing: 172.28.1.3:59484
06:21:12.747 DEBU minclient/conn:W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1: TCP connection established.
06:21:12.746 DEBU W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1_client: onDocument(): &{Epoch:116527 GenesisEpoch: 115586 SendRatePerMinute: 100 Mu: 0.001 MuMaxDelay: 9000 LambdaP:0.002 LambdaPMaxDelay:10000 LambdaL:0.0005 LambdaLMaxDelay:10000 LambdaD:0.0005 LambdaDMaxDelay:30000 LambdaM: 0.2 LambdaMMaxDelay: 100 SharedRandomValue: 8RIfUodR252GgoaTQlyCzhbM1UbCJj7Dpo91TzHuktU= PriorSharedRandom: [5TCLHt6+CANO6n5WCzu36dbzRUj+tITFrMxY0z/FwTE=, stL34EWNBD+sBt38J1uu1ZPPiM7l9ABcv+rC4bcYhdk=] Topology:[0]{&{Name:mix2 IdentityKey:1x8tq04hY+i83o9Yn2nrfTkFj4jXIMWCWilR7fgrWyM= LinkKey:yWxiHYt/4tYLumLi9GDXDnmTFg5d1aUxZ0huzVKLvX8= MixKeys:map[116527:bE9RmiZVrAGw+6KgAB/K/2ZwvFOjxFOD8lEXwXK+4EQ= 116528:34FNhoU5T1T01+k48qpHgnJVAXR1PQ+K2CcvSyT96wQ= 116529:aamqMKWqwad9GT+UUD/PZPqS0lhzCZsEoVOoBuSWsmw=] Addresses:map[tcp4:[172.28.1.6:29486]] Kaetzchen:map[] Layer:0 LoadWeight:0 AuthenticationType:}[1]{&{Name:mix1 IdentityKey:4r9jePAbuzhytcnM3nbD5UQOmzl1X1hI8QOMbhTrg8s= LinkKey:0XVV56EMIWhCHtXkw7xy/SSgbAfc40hx/bCN4CD6ETo= MixKeys:map[116527:b/WnxSgi4BWtcDhgOAPU0zX1FN2FQFOIHNtrUvzg0QM= 116528:iHR396QZsDTRVlwHdDkOMLdzw1N9xuAPl8nvNAlX3BU= 116529:sAOIzF+t43O50smHrinnc5s1b1N+x0rxtGOcB+44wnk=] Addresses:map[tcp4:[172.28.1.5:29485]] Kaetzchen:map[] Layer:1 LoadWeight:0 AuthenticationType:}},[2]{&{Name:mix3 IdentityKey:La2FnsbKoU8dzQhfN4zkxSYn6T7LRWmFn0JgtlEklQw= LinkKey:il0UH2I0LhQmb0up8oajH7Jm7LATqDnrWg4ZTugxKh0= MixKeys:map[116527:5a1eY+Nwxc7RMLMDmphAtRF+JfWfRv98Rosw0r+Isww= 116528:BEWJoKLOrHgrYfmhiaJHeZyrm1wQXM2ErJpOWa0sQCo= 116529:3sQAXotaaCMF6v/VCsOpzTZWZvaBAi3Ry6DXgBKD6ho=] Addresses:map[tcp4:[172.28.1.7:29487]] Kaetzchen:map[] Layer:2 LoadWeight:0 AuthenticationType:}},}, Providers:[]{&{Name:provider2 IdentityKey:imigzI26tTRXyYLXujLEPI9QrNYOEgC4DElsFdP9acQ= LinkKey:aTq4nyqV6GwsOEFl4zIxU3vZ3SjfSSCvLt8+04Is7CI= MixKeys:map[116527:qjP1Mr1AL74fRmJA27K5rrGIDnQvLQf/122Pw2Gti0s= 116528:UEldKfKBXRWnj5pBDgfAXbI/QaylkUutUBRPaF7GVF0= 116529:8un+m4RdMPxnfrP1cTATXVR8E22NDZ48kOWe6zK6Rzs=] Addresses:map[tcp4:[172.28.1.4:29484]] Kaetzchen:map[echo:map[endpoint:echo] reunion:map[endpoint:+reunion]] Layer:255 LoadWeight:0 AuthenticationType:oob},&{Name:provider1 IdentityKey:2krwfNDfbakZCSTUUZYKXwdduzlEgS9Jfwm7eyZ0sCg= LinkKey:bAAHyaf/fl682PBj2VWzGRMVfWBV5t+oBaaKUg+6ZhE= MixKeys:map[116527:zHRcV2dookTuQcImGMt8MSBVBXMUwLR/uM3IXmjoikU= 116528:VWqAJ3pm0KmQnyCBvq0qJvp9NVdubU+m5szwpe6vkAk= 116529:BgJ01aJqj0OoqVBukmNPPJmfaF5LpHgTOgaeV9h0iys=] Addresses:map[tcp4:[172.28.1.3:59484]] Kaetzchen:map[echo:map[endpoint:echo] panda:map[endpoint:+panda] spool:map[endpoint:+spool]] Layer:255 LoadWeight:0 AuthenticationType:tofu}}}
06:21:12.748 DEBU W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1_client: onDocument(): setting PollInterval to 320ms
06:21:12.753 DEBU W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1_client: SendMessage
06:21:12.753 DEBU catshadow: NewDocumentEvent: Setting readInboxTimer to 469ms
06:21:12.766 ERRO minclient/conn:W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1: Handshake failed: EOF
06:21:12.766 DEBU W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1_client: onConnection minclient/conn: connect error: EOF
06:21:12.767 DEBU minclient/conn:W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1: TCP connection closed.
06:21:12.767 DEBU minclient/conn:W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1: Connection terminated, will reconnect.
06:21:12.767 DEBU W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=@provider1_client: onConnection minclient/conn: not connected to the Provider
```

**Comment by mixmasala** [comment]
```
10:21:57.797 DEBU incoming:4: New incoming connection: 172.28.0.1:48254
10:21:57.850 ERRO provider: Authentication failed: User: 'w3mjy7glp+4yip6k0yfwfsc4gfis8fn+lcc9zrwcnru=', Key: 'W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU='
10:21:57.850 DEBU pki: Incoming: 'dzNtank3Z2xwKzR5aXA2azB5Zndmc2M0Z2ZpczhmbitsY2M5enJ3Y25ydT0=' AD not an IdentityKey?.
10:21:57.850 DEBU incoming:4: Authentication failed: 'dzNtank3Z2xwKzR5aXA2azB5Zndmc2M0Z2ZpczhmbitsY2M5enJ3Y25ydT0=' (W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=)
10:21:57.850 ERRO incoming:4: Handshake failed: wire/session: authentication failed
10:21:57.850 DEBU incoming:4: Closing.
10:22:42.875 DEBU listener:0: Accepted new connection: 172.28.0.1:48256
10:22:42.875 DEBU incoming:5: New incoming connection: 172.28.0.1:48256
10:22:42.925 ERRO provider: Authentication failed: User: 'w3mjy7glp+4yip6k0yfwfsc4gfis8fn+lcc9zrwcnru=', Key: 'W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU='
10:22:42.926 DEBU pki: Incoming: 'dzNtank3Z2xwKzR5aXA2azB5Zndmc2M0Z2ZpczhmbitsY2M5enJ3Y25ydT0=' AD not an IdentityKey?.
10:22:42.926 DEBU incoming:5: Authentication failed: 'dzNtank3Z2xwKzR5aXA2azB5Zndmc2M0Z2ZpczhmbitsY2M5enJ3Y25ydT0=' (W3MJY7GLp+4yIp6K0yFWFSC4gFIS8fN+lcc9zRWCNRU=)
10:22:42.926 ERRO incoming:5: Handshake failed: wire/session: authentication failed
10:22:42.926 DEBU incoming:5: Closing.
```

**Comment by mixmasala** [comment]
Note that I merged main into this branch and updated the configuration files

**Comment by mixmasala** [comment]
the strange thing is that in my configuration files, provider2 has TrustOnFirstUse set but is publishing its descriptor with the wrong authentication type (OutOfBandAuth), whereas provider1 has the opposite behavior.

**Comment by mixmasala** [comment]
the missing piece for this to work was for clients to accept a pki.Document iff it has at least one loop service present.

**Comment by mixmasala** [comment]
have we clearly articulated the model where the "entry" provider does not run a loop service?

File: client/worker.go [hunk 1]
```diff
@@ -225,11 +224,11 @@ func (s *Session) sendFromQueueOrDecoy(loopSvc *utils.ServiceDescriptor) {
 func (s *Session) isDocValid(doc *pki.Document) error {
 	for _, provider := range doc.Providers {
 		_, ok := provider.Kaetzchen[constants.LoopService]
-		if !ok {
-			return errors.New("found a Provider which does not have the loop service")
+		if ok {
```
Comment (mixmasala) [inline client/worker.go:227]: should the client require a particular fraction of loop services to service provider nodes, or require that if a provider has kaetzchen that it must also have a loop service?

**Review by mixmasala — COMMENTED** [review]
I added some changes to this branch, please see my questions inline.

File: client/worker.go (hunk 1 above)
Comment (david415) [reply client/worker.go:227]: that's out of scope for this feature addition

**Pull request closed without merging**